hypothesis-python/tests/numpy/test_gen_data.py

```diff
@@ -504,3 +504,212 @@ def test_minimize_negative_tuple_axes(ndim, data):
+
+
+@settings(deadline=None)
+@given(max_dim=st.integers(4, 6), data=st.data())
+def test_broadcastable_shape_adjusts_max_dim_with_explicit_bounds(max_dim, data):
+    # Ensures that `broadcastable_shapes` limits itself to satisfiable dimensions
+    # Broadcastable values can only be drawn for dims 0-3 for these shapes
+    shape = data.draw(st.sampled_from([(5, 3, 2, 1), (0, 3, 2, 1)]), label="shape")
+    broadcastable_shape = data.draw(
+        nps.broadcastable_shapes(
+            shape, min_side=2, max_side=3, min_dims=3, max_dims=max_dim
+        ),
+        label="broadcastable_shapes",
+    )
+    assert len(broadcastable_shape) == 3
+    a = np.zeros(shape)
+    b = np.zeros(broadcastable_shape)
+    np.broadcast(a, b)  # error if drawn shape for b is not broadcast-compatible
+
+
+@settings(deadline=None)
+@given(min_dim=st.integers(0, 4), min_side=st.integers(2, 3), data=st.data())
+def test_broadcastable_shape_shrinking_with_singleton_out_of_bounds(
+    min_dim, min_side, data
+):
+    max_dim = data.draw(st.one_of(st.none(), st.integers(min_dim, 4)), label="max_dim")
+    max_side = data.draw(
+        st.one_of(st.none(), st.integers(min_side, 6)), label="max_side"
+    )
+    ndims = data.draw(st.integers(1, 4), label="ndim")
+    shape = (1,) * ndims
+    smallest = minimal(
+        nps.broadcastable_shapes(
+            shape,
+            min_side=min_side,
+            max_side=max_side,
+            min_dims=min_dim,
+            max_dims=max_dim,
+        )
+    )
+    assert smallest == (min_side,) * min_dim
+
+
+@settings(deadline=None)
+@given(min_dim=st.integers(0, 4), min_side=st.integers(2, 3), data=st.data())
+def test_broadcastable_shape_shrinking_with_singleton_out_of_bounds(
+    min_dim, min_side, data
+):
+    max_dim = data.draw(st.one_of(st.none(), st.integers(min_dim, 4)), label="max_dim")
+    max_side = data.draw(
+        st.one_of(st.none(), st.integers(min_side, 6)), label="max_side"
+    )
+    ndims = data.draw(st.integers(1, 4), label="ndim")
+    shape = (1,) * ndims
+    smallest = minimal(
+        nps.broadcastable_shapes(
+            shape,
+            min_side=min_side,
+            max_side=max_side,
+            min_dims=min_dim,
+            max_dims=max_dim,
+        )
+    )
+    assert smallest == (min_side,) * min_dim
+
+
+@settings(deadline=None)
+@given(
+    shape=nps.array_shapes(min_dims=0, max_dims=3, min_side=0, max_side=5),
+    max_dims=st.integers(0, 6),
+    data=st.data(),
+)
+def test_broadcastable_shape_can_generate_arbitrary_ndims(shape, max_dims, data):
+    # ensures that generates shapes can possess any length in [min_dims, max_dims]
+    desired_ndim = data.draw(st.integers(0, max_dims), label="desired_ndim")
+    min_dims = data.draw(
+        st.one_of(st.none(), st.integers(0, desired_ndim)), label="min_dims"
+    )
+    args = (
+        dict(min_dims=min_dims) if min_dims is not None else {}
+    )  # check default arg behavior too
+    minimal(
```

Comment:
In this case, we don't need `minimal` - we just need to see a single witness that it's possible.  `find_any` will therefore be much more efficient :smile:

Reply:
Oh! Great call. That is much much better